src/lib/types/validation.ts

```diff
@@ -0,0 +1,165 @@
+  index: z.number(),
+  roleKey: role,
+})
+
+const inputHasMetadata = z.object({
+  index: z.number(),
+  metadataKey: tokenMetadataKey,
+})
+
+const matchInputOutputRole = z.object({
+  inputIndex: z.number(),
+  inputRoleKey: role,
+  outputIndex: z.number(),
+  outputRoleKey: role,
+})
+
+const matchInputOutputMetadataValue = z.object({
+  inputIndex: z.number(),
+  inputMetadataKey: tokenMetadataKey,
+  outputIndex: z.number(),
+  outputMetadataKey: tokenMetadataKey,
+})
+
+const matchInputIdOutputMetadataValue = z.object({
+  inputIndex: z.number(),
+  outputIndex: z.number(),
+  outputMetadataKey: tokenMetadataKey,
+})
+
+const fixedNumberOfInputs = z.object({
+  numInputs: z.number(),
+})
+
+const fixedNumberOfOutputs = z.object({
+  numOutputs: z.number(),
+})
+
+const fixedInputMetadataValue = z.object({
+  index: z.number(),
+  metadataKey: tokenMetadataKey,
+  metadataValue: metadataValue,
+})
+
+const fixedOutputMetadataValue = z.object({
+  index: z.number(),
+  metadataKey: tokenMetadataKey,
+  metadataValue: metadataValue,
+})
+
+const fixedOutputMetadataValueType = z.object({
+  index: z.number(),
+  metadataKey: tokenMetadataKey,
+  metadataValueType: metadataValueType,
+})
+
+export const restrictionValidation = z.union([
+  z.object({ None: none }).strict(),
+  z.object({ Fail: fail }).strict(),
+  z.object({ SenderHasInputRole: senderHasInputRole }).strict(),
+  z.object({ SenderHasOutputRole: senderHasOutputRole }).strict(),
+  z.object({ OutputHasRole: outputHasRole }).strict(),
+  z.object({ OutputHasMetadata: outputHasMetadata }).strict(),
+  z.object({ InputHasRole: inputHasRole }).strict(),
+  z.object({ InputHasMetadata: inputHasMetadata }).strict(),
+  z.object({ MatchInputOutputRole: matchInputOutputRole }).strict(),
+  z.object({ MatchInputOutputMetadataValue: matchInputOutputMetadataValue }).strict(),
+  z.object({ MatchInputIdOutputMetadataValue: matchInputIdOutputMetadataValue }).strict(),
+  z.object({ FixedNumberOfInputs: fixedNumberOfInputs }).strict(),
+  z.object({ FixedNumberOfOutputs: fixedNumberOfOutputs }).strict(),
+  z.object({ FixedInputMetadataValue: fixedInputMetadataValue }).strict(),
+  z.object({ FixedOutputMetadataValue: fixedOutputMetadataValue }).strict(),
+  z.object({ FixedOutputMetadataValueType: fixedOutputMetadataValueType }).strict(),
+])
+
+// leaving as it's being used in other file.
+export const stepValidation = z.union([
+  z.object({ op: binaryOperator }),
+  z.object({ restriction: restrictionValidation }),
+])
+
+export const programValidation = z.array(stepValidation)
+
+export const processValidation = z.object({
+  name: z.string().max(Constants.PROCESS_ID_LENGTH),
+  version: z.number(),
+  program: programValidation,
+})
+
+export const simpleProcesssValidation = z.array(z.object({ name: z.string() }).passthrough())
+
+export const processesValidation = z.array(processValidation)
+
+export type ValidationRestriction = z.infer<typeof restrictionValidation>
```

Comment:
Could argue about the naming, since this is Zod maybe say ZodProgramStep (less characters) but is just me trying to pick up on something so feel free to ignore.

Reply:
I think the aim should be to abstract `zod` away from our code so `Validation` is likely the right word here. A better thing would be to wrap and rethrow the `ZodError`.